src/rust/lib/shortcuts/src/lib.rs

```diff
@@ -0,0 +1,902 @@
+
+#[cfg(test)]
+mod benchmarks {
+    use super::*;
+    use test::Bencher;
+
+    const CONS_SIMPLE  : &'static str = "ctrl";
+    const CONS_COMPLEX : &'static str = "ctrl cmd alt shift";
+
+    // === Construction ===
+
+    #[bench]
+    fn automata_registry_construction_simple_without_optimization(bencher:&mut Bencher) {
+        construction::<AutomataRegistry<i32>>(CONS_SIMPLE,false,bencher);
+    }
+
+    #[bench]
+    fn automata_registry_construction_complex_without_optimization(bencher:&mut Bencher) {
+        construction::<AutomataRegistry<i32>>(CONS_COMPLEX,false,bencher);
+    }
+
+    #[bench]
+    fn automata_registry_construction_simple_with_optimization(bencher:&mut Bencher) {
+        construction::<AutomataRegistry<i32>>(CONS_SIMPLE,true,bencher);
+    }
+
+    #[bench]
+    fn automata_registry_construction_complex_with_optimization(bencher:&mut Bencher) {
+        construction::<AutomataRegistry<i32>>(CONS_COMPLEX,true,bencher);
+    }
+
+    #[bench]
+    fn hashset_registry_construction_simple(bencher:&mut Bencher) {
+        construction::<HashSetRegistry<i32>>(CONS_SIMPLE,true,bencher);
+    }
+
+    #[bench]
+    fn hashset_registry_construction_complex(bencher:&mut Bencher) {
+        construction::<HashSetRegistry<i32>>(CONS_COMPLEX,true,bencher);
+    }
+
+    fn construction<T:Registry<i32>>(input:&str, optimize:bool, bencher:&mut Bencher) -> T {
+        bencher.iter(|| {
+            let mut registry : T = default();
+            let max_count        = test::black_box(10);
+            for i in 0..max_count {
+                registry.add(Press,format!("{} a{}",i,input),i);
+            }
+            if optimize {
+                registry.optimize();
+            }
+        });
+        default()
+    }
+
+
+    // === Lookup ===
+
+    #[bench]
+    fn automata_registry_lookup(bencher:&mut Bencher) {
+        lookup::<AutomataRegistry<i32>>(bencher);
+    }
+
+    #[bench]
+    fn hashset_registry_lookup(bencher:&mut Bencher) {
+        lookup::<HashSetRegistry<i32>>(bencher);
+    }
+
+    fn lookup<T:Registry<i32>>(bencher:&mut Bencher) -> T {
+        let mut registry : T = default();
+        let nothing          = Vec::<i32>::new();
+        let max_count        = test::black_box(100);
+        for i in 0..max_count {
+            registry.add(Press, format!("ctrl shift a{}",i), i);
+        }
+        registry.optimize();
+        bencher.iter(|| {
+            for i in 0..max_count {
+                let key = format!("a{}",i);
+                assert_eq!(registry.on_press("ctrl-left"),nothing);
+                assert_eq!(registry.on_press("shift-left"),nothing);
+                assert_eq!(registry.on_press(&key),vec![i]);
+                assert_eq!(registry.on_release(&key),nothing);
+                assert_eq!(registry.on_release("shift-left"),nothing);
+                assert_eq!(registry.on_release("ctrl-left"),nothing);
+            }
+        });
+        registry
+    }
+}
```

Comment:
I'd leave it for the next PRs. I'd rather keep it in one file now, as it might change during development of focus management and I want to kee it close to each other.